### PR TITLE
Hessian-vector product related to 

### PR DESCRIPTION
**Pull request by kliegeois**
This PR covers several improvements:

1. The Hessian-vector product computations related to scalar parameters are added both for responses and residuals,
2. Solves #663,
3. Solves a bug related to the zero-out of the Lagrange multipliers associated to Dirichlet boundary conditions when using more than 1 MPI process; this is now tested in PyAlbany using the steadyHeat.py python test,
4. An unused option of the steadyHeat.py python test has been removed,
5. It is now possible to specify the AD support for the Hessian-vector product differently for each contribution in the yaml file; some conditions are required to have meaningful computations, if one of those is not verified, an error is thrown explaining what should be updated in the .yaml file.

This PR has been tested on Blake with and without `MESH_DEPENDS_ON_PARAMETERS` enabled (with [PR #8806](https://github.com/trilinos/Trilinos/pull/8806)).
The code builds _warning free_, i.e., no errors are thrown while building with warning treated as errors with the previously defined considered set of warnings.

@mperego @ikalash

### CODE REVIEW

File: src/evaluators/utility/PHAL_Source_Def.hpp [hunk 1]
```diff
@@ -17,6 +17,9 @@
 #include "Teuchos_Array.hpp"
 #include "Teuchos_TestForException.hpp"
 
+#include "PHAL_SharedParameter.hpp"
+#include "PHAL_Spatial.hpp"
```
Comment (mperego) [inline src/evaluators/utility/PHAL_Source_Def.hpp:21]: The name `PHAL_Spatial` is a bit obsucre. We could call it `PHAL_Field_Source` or `Source Functions`

**Review by mperego — APPROVED** [review]
@kliegeois  A lot of work!
I didn't really try to understand every single line of code, but overall looks great.
I added a couple of comments. They are just suggestions, no need to modify them.
